README.md
# DNAQuery

[![Go Report Card](https://goreportcard.com/badge/github.com/lifechurch/dnaquery)](https://goreportcard.com/report/github.com/lifechurch/dnaquery) [![Build Status](https://travis-ci.org/lifechurch/dnaquery.svg?branch=master)](https://travis-ci.org/lifechurch/dnaquery)
[![Coverage Status](https://coveralls.io/repos/github/lifechurch/dnaquery/badge.svg?branch=master)](https://coveralls.io/github/lifechurch/dnaquery?branch=master)

DNAQuery is a command line utility to take [LogDNA](https://logdna.com) archives and load them into [BigQuery](https://cloud.google.com/bigquery/). This allows long-term, queryable storage of logs in BigQuery (which is considerably more cost effective). In our use case, we want real-time access to many different logs and long-term storage of a subset of those logs. This approach has helped us find a balance between access and cost.

## Getting Started

`go get -u github.com/lifechurch/dnaquery`

DNAQuery has been tested on Go v1.11

## Prerequisites

- a LogDNA account with [GCS Archiving](https://docs.logdna.com/docs/archiving#section-google-cloud-storage) enabled
- a [Google Cloud Platform](https://cloud.google.com) account

## Configuring

`cp example.toml dnaquery.toml`

Edit dnaquery.toml

All of these settings are currently required.

### Storage

```
[storage]
LogDirectory = 'logs' # temp local directory to download logs to and store results, this directory will be created if it doesn't exist
```

### Apps
```
[[apps]]
    Name = "production-app"  # name of app
    Regex = '^([\d.]+) - \[([^\]]*)\] - - \[([^\]]*)\] "([^"]*)" (\d+) (\d+) "([^"]*)" "([^"]*)" (\d+) ([\d.]+)  ([\d.:]+) (\d+) ([\d.]+) (\d+)$'  # regex used to pull parts of logs out, currently we don't use any named capture groups
    TimeGroup = 3 # the number of the capture group (1-based) in above regex that holds the time component of the log
    TimeFormat = "2/Jan/2006:15:04:05 -0700" # the format of the time field using https://golang.org/pkg/time/#Parse
    [[apps.excludes]]  # an array of tables for exclusions
        Group = 4  # the number of the capture group in above regex to be used in the exclusion check
        Contains = "ping" # excludes logs if the value here is contained in the string in the above capture group
```

### GCP
```
[gcp]  # Google Cloud Project settings
ProjectID = "gcpproj" # name of GCP project
CredentialsFile = "gcp_credentials.json" # relative or absolute path to the credentials file downloaded from GCP
Bucket = "logdna_to_bq" # name of bucket in Google Cloud Storage to save results for ingestion into BigQuery, bucket will need to be created before first run
Dataset = "logdna" # BigQuery dataset
TemplateTable = "logdna" # currently DNAQuery uses a template table. More details below.
LogBucket = "logs" # name of bucket set in LogDNA archive setup above
LogPrefix = "a7112abc9d" # each archive file starts with a prefix specific to your LogDNA account
```

#### TemplateTable

Currently the schema for the BigQuery table is specified by creating a table with the schema that matches the regex. This is likely to change in the future as we may need a different schema for each app above.

## Usage

`dnaquery --date 2017-11-20`

## Dependencies

This project uses [dep](https://github.com/golang/dep) for dependency management

```
go get -u github.com/golang/dep/cmd/dep
dep ensure -update
```

## testing

```
go test -v ./...
```

## Changelog

[CHANGELOG.md](CHANGELOG.md)

## Contributing

Please read [CONTRIBUTING.md](CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests.

## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/lifechurch/dnaquery/tags).

## Authors

* **Chris Vaughn** - *Initial work* - [chrisvaughn](https://github.com/chrisvaughn)

See also the list of [contributors](https://github.com/lifechurch/dnaquery/contributors) who participated in this project.

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Changes
- update to go modules 

## [0.3.1] - 2018-04-23
### Changes
- remove `_` from BigQuery table names to comply with BigQuery recommendations.

## [0.3.0] - 2018-03-09
### Changes
- `containers` in the config file is renamed to `apps` in order to be more flexible

## [0.2.0] - 2018-02-15
### Added
- clean up local log files when job is complete
- added CLI framework for easier handling of options
- added tests from 0 to 31% test coverage
- use LogDNA archive files in Google Cloud Storage instead of AWS S3 for bandwidth cost savings

### Changed
- `Bucket` in GCP config is now `UploadBucket`

### Removed
- AWS S3 config and download of log archive from AWS. Check latest README for up to date config documentation 

## [0.1.0] - 2017-11-12
### Added
- Working code to take LogDNA archives in AWS S3, process them, and ingest them BigQuery. 

[Unreleased]: https://github.com/lifechurch/dnaquery/compare/v0.3.1...HEAD
[0.3.1]: https://github.com/lifechurch/dnaquery/compare/v0.3.0...v0.3.1
[0.3.0]: https://github.com/lifechurch/dnaquery/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/lifechurch/dnaquery/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/lifechurch/dnaquery/compare/9bece45e6dfdb371b54a765a672429f9958bc2ca...v0.1.0


main_test.go
package main

import (
	"io/ioutil"
	"os"
	"strings"
	"testing"
	"time"
)

func TestNewDNAQuery(t *testing.T) {
	_, err := NewDNAQuery(&Configuration{})
	if err == nil {
		t.Errorf("should error if 0 Apps in config")
	}
	c1 := App{Name: "App 1"}
	cfg := &Configuration{
		Apps: []App{c1},
	}
	_, err = NewDNAQuery(cfg)
	if err == nil {
		t.Errorf("should error if no log directory is specified")
	}

	cfg, _ = NewConfiguration("example.toml")
	dna, _ := NewDNAQuery(cfg)
	if dna == nil {
		t.Error("dna should not be nil")
	}

}

func TestCleanupFiles(t *testing.T) {
	// make a few files
	f1Path := "file1"
	f2Path := "file2"
	f3Path := "file3" // this file is never created
	f1, err := os.Create(f1Path)
	if err != nil {
		t.Fatalf("Couldn't create test file (%s): %s", f1Path, err.Error())
	}
	f1.Close()
	f2, err := os.Create(f2Path)
	if err != nil {
		t.Fatalf("Couldn't create test file (%s): %s", f1Path, err.Error())
	}
	f2.Close()
	cleanupFiles(f1Path, f2Path, f3Path)

	if _, err := os.Stat(f1Path); !os.IsNotExist(err) {
		t.Errorf("test file (%s) still exists", f1Path)
	}
	if _, err := os.Stat(f2Path); !os.IsNotExist(err) {
		t.Errorf("test file (%s) still exists", f2Path)
	}
}

func TestProcessLine(t *testing.T) {
	c1 := App{
		Name:       "app1",
		Regex:      `^([\d.]+) \[([^\]]*)\] - "([^"]*)" (\d+)`,
		TimeGroup:  2,
		TimeFormat: "2/Jan/2006:15:04:05 -0700",
	}
	c2 := App{
		Name:       "app2",
		Regex:      `^([\d.]+) - \[([^\]]*)\] - "([^"]*)" (\d+)`,
		TimeGroup:  2,
		TimeFormat: "2006-01-02T15:04:05-0700",
		Excludes:   []Exclude{{Group: 3, Contains: "ping"}},
	}
	// invalid configuration, should have code to detect this at start up, for now
	// make sure code handles it
	c3 := App{
		Name:       "app3",
		Regex:      `^([\d.]+)`,
		TimeGroup:  2,
		TimeFormat: "2006-01-02T15:04:05-0700",
		Excludes:   []Exclude{{Group: 3, Contains: "ping"}},
	}
	cfg := &Configuration{
		Apps:    []App{c1, c2, c3},
		Storage: Storage{LogDirectory: "/tmp/"},
	}
	dna, err := NewDNAQuery(cfg)
	if err != nil {
		t.Fatalf("unexpected error %v", err)
	}
	err = dna.processLine("", nil)
	if err == nil {
		t.Fatal("Expected error for bad filename")
	}

	ch := make(chan [2]string)
	outfile := "output.csv"
	go func(outfile string, ch chan [2]string) {
		// regular, expect normal operation
		ch <- [2]string{"app1", `123.123.123.123 [13/Nov/2017:13:23:01 -0000] - "GET view.json" 200`}
		ch <- [2]string{"app1", `123.123.123.123 [13/Nov/2017:13:23:04 -0000] - "GET ping.json" 200`}
		ch <- [2]string{"app2", `2.1.5.3 - [2017-12-03T13:23:04-0500] - "GET ping.json" 200`}
		ch <- [2]string{"app2", `2.1.5.3 - [2017-12-03T13:23:04-0500] - "GET view.json" 200`}
		// case where exclusion and time groups are more then regex
		ch <- [2]string{"app3", `2.1.5.3`}
		// case where there is no app registered
		ch <- [2]string{"app", `2.1.5.3 - [2017-12-03T13:23:04-0500] - "GET view.json" 200`}
		// case where the log line doesn't match regex
		ch <- [2]string{"app1", `error`}
		// case where time format does not match, should just skip over
		ch <- [2]string{"app1", `123.123.123.123 [13-Nov-2017:13:23:01 -0000] - "GET view.json" 200`}
		close(ch)
	}(outfile, ch)
	err = dna.processLine(outfile, ch)
	// sleep a bit for goroutine to finish up once channel is closed
	time.Sleep(500 * time.Millisecond)
	dat, err := ioutil.ReadFile(outfile)
	if err != nil {
		t.Fatalf("unexpected error opening outfile, %v", err)
	}
	lines := strings.Split(string(dat), "\n")
	expectedLines := 5 // 4 logs + 1 empty line
	if len(lines) != expectedLines {
		t.Errorf("expected %d lines, found %d lines", expectedLines, len(lines))
	}
	cleanupFiles(outfile)
}

func TestReadLine(t *testing.T) {
	c1 := App{
		Name:  "ingress-lb",
		Regex: `^.*$`,
	}
	cfg := &Configuration{
		Apps:    []App{c1},
		Storage: Storage{LogDirectory: "/tmp/"},
	}
	dna, err := NewDNAQuery(cfg)
	if err != nil {
		t.Fatalf("unexpected error %v", err)
	}
	ch, err := dna.readLine("test_data/testlog")
	if err == nil {
		t.Error("expected error for invalid file")
	}
	ch, err = dna.readLine("main.go")
	if err == nil {
		t.Error("expected error for invalid gzip file")
	}

	ch, err = dna.readLine("test_data/testlog.json.gz")
	if err != nil {
		t.Fatalf("unexpected error in readLine, %v", err)
	}
	count := 0
	for _ = range ch {
		count += 1
	}
	if count != 6 {
		t.Errorf("Unexpected # of logs found based on config, expected 6 found %v", count)
	}
}


config_test.go
package main

import (
	"crypto/rand"
	"os"
	"testing"

	"encoding/hex"
	"path/filepath"

	"github.com/google/go-cmp/cmp"
)

func TestReadConfig(t *testing.T) {
	_, err := NewConfiguration("doesnotexist.toml")

	if err == nil {
		t.Error("should have returned an error")
	}
	_, err = NewConfiguration("example.toml")
	if err != nil {
		t.Error("could not parse toml file when should have", err.Error())
	}
	// create bad file
	f := "bad.toml"
	f1, _ := os.Create(f)
	f1.Write([]byte("bad toml file"))
	f1.Close()
	_, err = NewConfiguration(f)
	if err == nil {
		t.Errorf("should have error reading bad toml file")
	}
	// clean up test file
	os.Remove(f)
}

func TestGetApp(t *testing.T) {
	// build a config with 1 App
	a1 := App{Name: "App 1"}
	cfg := &Configuration{
		Apps: []App{a1},
	}
	_, err := cfg.getApp("App That Doesn't Exist")
	if err == nil {
		t.Errorf("Err not returned when app doesn't exist")
	}

	a1Retrieved, _ := cfg.getApp("App 1")
	if !cmp.Equal(a1, a1Retrieved) {
		t.Errorf("a1 was not equal")
	}
}

func TestCompileRegex(t *testing.T) {
	a1 := App{
		Name:  "App 1",
		Regex: `([\d.]+) - \[([^\]]*)\] - - \[([^\]]*)\]`,
	}
	cfg := &Configuration{
		Apps: []App{a1},
	}
	if cfg.Apps[0].CompiledRegex != nil {
		t.Errorf("compile regex is not nil but compile hasn't been called yet")
	}
	cfg.compileRegex()
	if cfg.Apps[0].CompiledRegex == nil {
		t.Errorf("compile regex is nil")
	}
}

func TestSetupDir(t *testing.T) {
	randBytes := make([]byte, 16)
	rand.Read(randBytes)

	path := filepath.Join(os.TempDir(), hex.EncodeToString(randBytes))
	cfg := &Configuration{
		Storage: Storage{LogDirectory: path},
	}
	err := cfg.setupDirectory()
	if err != nil {
		t.Errorf("directory setup failed: %v", err)
	}
	_, err = os.Stat(path)
	if os.IsNotExist(err) {
		t.Errorf("directory does not exist: %v", err)
	}
	os.Remove(path)

	path = ""
	cfg = &Configuration{
		Storage: Storage{LogDirectory: path},
	}
	err = cfg.setupDirectory()
	if err == nil {
		t.Error("should have error creating empty dir")
	}
}

func TestExtractAppNames(t *testing.T) {
	cfg := &Configuration{
		Apps: []App{{Name: "a1"}, {Name: "a2"}},
	}
	names := cfg.extractAppNames()
	expected := make(map[string]struct{}, len(cfg.Apps))
	expected["a1"] = struct{}{}
	expected["a2"] = struct{}{}
	if !cmp.Equal(names, expected) {
		t.Errorf("expected names to equal %v, received %v instead", expected, names)
	}
}


config.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"regexp"

	"strings"

	"github.com/pelletier/go-toml"
	"github.com/pkg/errors"
)

// Storage holds the configuration for [storage] section of the toml config.
type Storage struct {
	LogDirectory string
}

// GCP holds the configuration for [gcp] section of the toml config.
type GCP struct {
	ProjectID       string
	CredentialsFile string
	UploadBucket    string
	Dataset         string
	TemplateTable   string
	LogPrefix       string
	LogBucket       string
}

// Exclude holds the configuration for the [[apps.excludes]] subsection
// of the toml config.
type Exclude struct {
	Group    int
	Contains string
}

// App holds the configuration for a single entry in the [[apps]]
// section of the toml config.
type App struct {
	Name          string
	Regex         string
	CompiledRegex *regexp.Regexp
	TimeGroup     int
	TimeFormat    string
	Excludes      []Exclude
}

func (app *App) isExcluded(r []string) bool {
	exclude := false
	for _, e := range app.Excludes {
		if e.Group >= len(r) {
			log.Printf("skipping exclusion: %v, Group not found in result", e)
			continue
		}
		if strings.Contains(r[e.Group], e.Contains) {
			exclude = true
			break
		}
	}
	return exclude
}

// Configuration holds the full configuration loaded from the toml config file.
type Configuration struct {
	Storage Storage
	GCP     GCP
	Apps    []App
}

func (cfg *Configuration) getApp(c string) (App, error) {
	for _, app := range cfg.Apps {
		if c == app.Name {
			return app, nil
		}
	}
	return App{}, errors.New("App not found")
}

func (cfg *Configuration) extractAppNames() (set map[string]struct{}) {
	set = make(map[string]struct{}, len(cfg.Apps))
	for _, app := range cfg.Apps {
		set[app.Name] = struct{}{}
	}
	return set
}

func (cfg *Configuration) compileRegex() {
	for i, c := range cfg.Apps {
		cmp := regexp.MustCompile(c.Regex)
		cfg.Apps[i].CompiledRegex = cmp
	}
}

func (cfg *Configuration) setupDirectory() error {
	err := os.MkdirAll(cfg.Storage.LogDirectory, os.ModePerm)
	if err != nil {
		err = errors.Wrap(err, fmt.Sprintf("Unable to create dir %s", cfg.Storage.LogDirectory))
	}
	return err
}

// NewConfiguration takes a path to a toml file and returns a new Configuration
func NewConfiguration(path string) (*Configuration, error) {
	cfg := &Configuration{}
	data, err := ioutil.ReadFile(path)
	if err != nil {
		return nil, errors.Wrap(err, fmt.Sprintf("Unable to open config (%s)", path))
	}
	err = toml.Unmarshal(data, cfg)
	if err != nil {
		return nil, errors.Wrap(err, "Error loading config")
	}
	return cfg, nil
}


main.go
package main

import (
	"bufio"
	"compress/gzip"
	"context"
	"encoding/csv"
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"
	"strings"
	"time"

	"cloud.google.com/go/bigquery"
	"cloud.google.com/go/storage"
	"github.com/buger/jsonparser"
	"github.com/pkg/errors"
	"github.com/urfave/cli"
	"google.golang.org/api/option"
)

var version = "0.3.1"

// DNAQuery holds config and derived data
type DNAQuery struct {
	*Configuration
	appNames map[string]struct{}
}

// NewDNAQuery returns a DNAQuery instance all setup complete
func NewDNAQuery(cfg *Configuration) (*DNAQuery, error) {
	if len(cfg.Apps) < 1 {
		return nil, errors.New("Configuration needs at least 1 app")
	}

	dna := &DNAQuery{
		Configuration: cfg,
		appNames:      cfg.extractAppNames(),
	}
	dna.compileRegex()
	err := dna.setupDirectory()
	if err != nil {
		return nil, errors.Wrap(err, "Error setting up directory")
	}
	return dna, nil
}

func cleanupFiles(path ...string) {
	for _, f := range path {
		err := os.Remove(f)
		if err != nil {
			log.Printf("Unable to delete file (%s): %s", f, err.Error())
		}
	}
}

func (d *DNAQuery) readLine(path string) (chan [2]string, error) {
	log.Println("Opening Logfile", path)
	inFile, err := os.Open(path)
	if err != nil {
		return nil, errors.Wrapf(err, "Unable to open input: %s", path)
	}
	gzReader, err := gzip.NewReader(inFile)
	if err != nil {
		return nil, errors.Wrap(err, "Unable to create gzip reader")
	}
	ch := make(chan [2]string, 50)
	go func(inFile io.ReadCloser, ch chan [2]string) {
		defer close(ch)
		defer inFile.Close()

		scanner := bufio.NewScanner(inFile)
		log.Println("Scanning log file")
		lineCount := 0
		for scanner.Scan() {
			lineCount++
			data := scanner.Bytes()
			app, _ := jsonparser.GetString(data, "_app")
			if _, ok := d.appNames[app]; ok {
				line, _ := jsonparser.GetString(data, "_line")
				ch <- [2]string{app, line}
			}
		}
		if err := scanner.Err(); err != nil {
			log.Fatalf("Error reading log file: %v", err)
		}
		log.Printf("Scanning complete. %d lines scanned\n", lineCount)
	}(gzReader, ch)
	return ch, nil
}

func (d *DNAQuery) processLine(path string, ch chan [2]string) error {
	log.Println("Starting processLine")
	csvOut, err := os.Create(path)
	if err != nil {
		return errors.Wrapf(err, "Unable to open output: %s", path)
	}
	w := csv.NewWriter(csvOut)
	defer csvOut.Close()
	nMatches := 0
	nSkipped := 0
	for r := range ch {
		app := r[0]
		line := r[1]
		var record []string
		record = append(record, app)
		a, err := d.getApp(app)
		if err != nil {
			log.Println("Can't find app config in processLine:", app)
			continue
		}
		result := a.CompiledRegex.FindStringSubmatch(line)
		if len(result) == 0 {
			nSkipped++
			continue
		}
		// check exclusion rules
		if a.isExcluded(result) {
			nSkipped++
			continue
		}
		// change time format
		if a.TimeGroup >= len(result) {
			log.Println("skipping time formating. Group not found in result")
		} else {
			dt, err := time.Parse(a.TimeFormat, result[a.TimeGroup])
			if err != nil {
				nSkipped++
				continue
			}
			result[a.TimeGroup] = dt.Format("2006-01-02 15:04:05 -07:00")
		}
		nMatches++
		record = append(record, result[1:]...)
		if err = w.Write(record); err != nil {
			return errors.Wrapf(err, "Unable to write record: %v", record)
		}
	}
	// Write any buffered data.
	w.Flush()

	if err := w.Error(); err != nil {
		log.Fatalln("Error creates csv", err.Error())
		return errors.Wrap(err, "Unable to create csv")
	}
	log.Printf("Matched %d lines, Skipped %d lines\n", nMatches, nSkipped)
	log.Println("Completed processLine")

	return nil
}

func (d *DNAQuery) getLogfile(logDate string) (logName string) {
	logName = d.GCP.LogPrefix + "." + logDate + ".json.gz"
	localLogName := filepath.Join(d.Storage.LogDirectory, logName)

	os.Setenv("GOOGLE_CLOUD_PROJECT", d.GCP.ProjectID)
	ctx := context.Background()
	client, err := storage.NewClient(ctx, option.WithCredentialsFile(d.GCP.CredentialsFile))
	CheckErr("Error creating storage client: ", err)
	bkt := client.Bucket(d.GCP.LogBucket)

	obj := bkt.Object(logName)
	objAttrs, err := obj.Attrs(ctx)
	CheckErr("Error getting object attrs: ", err)
	sizeInGCS := objAttrs.Size
	log.Printf("File in GCS is %f GB", float64(sizeInGCS)/1024/1024/1024)

	downloadFile := true
	if file, err := os.Open(localLogName); err == nil {
		stat, _ := file.Stat()
		// if file on disk is same size don't download
		downloadFile = stat.Size() != sizeInGCS
		file.Close()
	}

	if downloadFile {
		log.Printf("Downloading from GCS")

		file, err := os.Create(localLogName)
		if err != nil {
			log.Fatalf("Unable to open file %q, %v", logName, err)
		}
		defer file.Close()

		r, err := obj.NewReader(ctx)
		CheckErr("Error getting object reader: ", err)
		defer r.Close()
		numBytes, err := io.Copy(file, r)
		if err != nil {
			log.Fatalf("Unable to download item %q, %v", logName, err)
		}
		fmt.Println("Downloaded", file.Name(), numBytes, "bytes")
	} else {
		log.Printf("Skipping Download")
	}
	return localLogName
}

func (d *DNAQuery) uploadToGCS(path string, object string) error {
	log.Println("Starting upload to GCS")
	os.Setenv("GOOGLE_CLOUD_PROJECT", d.GCP.ProjectID)
	ctx := context.Background()
	client, err := storage.NewClient(ctx, option.WithCredentialsFile(d.GCP.CredentialsFile))
	CheckErr("Error creating storage client: ", err)
	f, err := os.Open(path)
	if err != nil {
		return err
	}
	defer f.Close()

	stat, _ := f.Stat()
	log.Printf("Upload size: %f MB\n", float64(stat.Size())/1024/1024)
	wc := client.Bucket(d.GCP.UploadBucket).Object(object).NewWriter(ctx)
	nBytes, nChunks := int64(0), int64(0)
	r := bufio.NewReader(f)
	buf := make([]byte, 0, 4*1024)
	for {
		n, err := r.Read(buf[:cap(buf)])
		buf = buf[:n]
		if n == 0 {
			if err == nil {
				continue
			}
			if err == io.EOF {
				break
			}
			log.Fatal(err)
		}
		nChunks++
		nBytes += int64(len(buf))
		_, err = wc.Write(buf)
		if err != nil {
			return err
		}
		if err != nil && err != io.EOF {
			log.Fatal(err)
		}
	}

	if err := wc.Close(); err != nil {
		return err
	}
	if err := client.Close(); err != nil {
		log.Fatal("Error closing GCS:", err.Error())
	}
	log.Println("Completed upload to GCS")
	return nil
}

func (d *DNAQuery) loadInBQ(object string, date string) {
	date = strings.Replace(date, "-", "", -1)
	log.Println("Starting load into BQ")
	ctx := context.Background()

	client, err := bigquery.NewClient(ctx, d.GCP.ProjectID,
		option.WithCredentialsFile(d.GCP.CredentialsFile))
	CheckErr("Error creating BQ Client: ", err)
	myDataset := client.Dataset(d.GCP.Dataset)

	templateTable := myDataset.Table(d.GCP.TemplateTable)

	gscURL := "gs://" + d.GCP.UploadBucket + "/" + object
	gcsRef := bigquery.NewGCSReference(gscURL)
	tmpTableMeta, err := templateTable.Metadata(ctx)
	CheckErr("Error getting template table: ", err)
	gcsRef.Schema = tmpTableMeta.Schema
	tableName := date
	loader := myDataset.Table(tableName).LoaderFrom(gcsRef)
	loader.CreateDisposition = bigquery.CreateIfNeeded
	loader.WriteDisposition = bigquery.WriteTruncate

	job, err := loader.Run(ctx)
	CheckErr("Error running job: ", err)
	log.Println("BQ Job created...")
	pollInterval := 5 * time.Second
	for {
		status, err := job.Status(ctx)
		CheckErr("Job Error: ", err)
		if status.Done() {
			if status.Err() != nil {
				log.Fatalf("Job failed with error %v", status.Err())
			}
			break
		}
		time.Sleep(pollInterval)
	}
	log.Println("Completed load into BQ")
}

func run(c *cli.Context) error {
	dateToProcess := c.String("date")
	if dateToProcess == "" {
		fmt.Print("ERROR: --date required\n\n")
		cli.ShowAppHelp(c)
		return errors.New("Error: --date required")
	}
	configFile := c.String("config")

	// load config
	cfg, err := NewConfiguration(configFile)
	CheckErr("Error: ", err)

	dna, err := NewDNAQuery(cfg)
	CheckErr("Error: ", err)

	logName := dna.getLogfile(dateToProcess)
	if !c.Bool("keep") {
		defer cleanupFiles(logName)
	}

	ch, err := dna.readLine(logName)
	CheckErr("Error: ", err)

	outFile := "results_" + dateToProcess + ".csv"
	outPath := filepath.Join(cfg.Storage.LogDirectory, outFile)
	if !c.Bool("keep") {
		defer cleanupFiles(outPath)
	}
	err = dna.processLine(outPath, ch)
	if err != nil {
		return errors.Wrap(err, "Error in processLine")
	}

	gcsObject := dateToProcess + "_results.csv"
	err = dna.uploadToGCS(outPath, gcsObject)
	CheckErr("Error uploading to GCS:", err)

	dna.loadInBQ(gcsObject, dateToProcess)

	if c.Bool("keep") {
		log.Printf("Log file not removed: %s", logName)
		log.Printf("Results file not removed: %s", outPath)
	}

	return nil
}

func main() {
	app := cli.NewApp()

	app.Flags = []cli.Flag{
		cli.StringFlag{
			Name:  "config, c",
			Usage: "Load configuration from `FILE`",
			Value: "dnaquery.toml",
		},
		cli.StringFlag{
			Name:  "date, d",
			Usage: "process log archive for `YYYY-MM-DD`",
		},
		cli.BoolFlag{
			Name:  "keep, k",
			Usage: "flag to keep log files after run, helpful during configuration",
		},
	}
	app.Action = run
	app.Version = version
	app.Run(os.Args)
}
